In [1]:
import pandas as pd
import numpy as np
import datetime
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn import tree
import seaborn as sns
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
import tensorflow as tf
from sklearn.metrics import mean_squared_error
import keras_tuner as kt
from sklearn import metrics

In [2]:
df_insurance_claims = pd.read_csv("D:/Postgraduate_studies_BA/Term_2/Applied_Deep_Learning/Week5/Insurance_claims.csv")

## Model 1: The Simple model based on the incident occurs within t days of the subscription date.
### Data Pre-processing

In [3]:
df_insurance_claims.head()

ReferenceId  PolicyholderNumber  FirstPartyVehicleNumber  \
0         4842              531112                 715507.0   
1         4844               87170                  71164.0   
2         4848               98706                 442609.0   
3         4849               38240                  24604.0   
4         4850               11339                   2933.0   

   ThirdPartyVehicleNumber                                       InsurerNotes  \
0                      NaN  avoids a cat and hits a garage pole With deduc...   
1                      NaN  accident only expert contacts us to inform us ...   
2                      NaN  ae Miss/ for garage change A/ setting up EAD/ ...   
3                      NaN  awaiting report to determine rc, no box checke...   
4                 229134.0  Insured in THIRD-PARTY formula Insured in a su...   

  PolicyholderOccupation LossDate FirstPolicySubscriptionDate  \
0           CivilServant   1/2/19                     6/18/18   
1                 Worker   1/2/19                     6/29/17   
2                 Worker   1/2/19                      2/5/17   
3           CivilServant   1/2/19                     1/21/17   
4                 Farmer   1/2/19                     1/13/18   

                         ClaimCause  \
0               CollisionWithAnimal   
1                     LossOfControl   
2  AccidentWithIdentifiedThirdParty   
3  AccidentWithIdentifiedThirdParty   
4  AccidentWithIdentifiedThirdParty   

                               ClaimInvolvedCovers  ...  \
0                     MaterialDamages ActLiability  ...   
1                     MaterialDamages ActLiability  ...   
2                     MaterialDamages ActLiability  ...   
3  MaterialDamages ActLiability ReplacementVehicle  ...   
4                                     ActLiability  ...   

  NumberOfPoliciesOfPolicyholder FpVehicleAgeMonths EasinessToStage  \
0                              1              104.0            0.25   
1                              3              230.0            0.50   
2                              9               93.0            0.25   
3                              2               56.0            0.25   
4                              4              110.0            0.25   

  ClaimWihoutIdentifiedThirdParty ClaimAmount  LossHour  PolicyHolderAge  \
0                               1     4624.73       8.0             45.0   
1                               1     1606.81      11.0             20.0   
2                               0      998.20      18.0             32.0   
3                               0     2506.92      11.0             46.0   
4                               0       12.00      12.0             28.0   

   NumberOfBodilyInjuries  FirstPartyLiability  Fraud  
0                       0                  1.0      0  
1                       0                  1.0      0  
2                       0                  0.5      0  
3                       0                  0.5      0  
4                       0                  0.0      0  

[5 rows x 26 columns]

In [4]:
# Check the missing values for each column
df_insurance_claims.isna().sum()

ReferenceId                            0
PolicyholderNumber                     0
FirstPartyVehicleNumber              495
ThirdPartyVehicleNumber            11151
InsurerNotes                        2357
PolicyholderOccupation               343
LossDate                               0
FirstPolicySubscriptionDate            0
ClaimCause                           197
ClaimInvolvedCovers                  195
DamageImportance                   10792
FirstPartyVehicleType                 12
ConnectionBetweenParties           11432
LossPostCode                        1408
PolicyHolderPostCode                   0
PolicyWasSubscribedOnInternet          0
NumberOfPoliciesOfPolicyholder         0
FpVehicleAgeMonths                    12
EasinessToStage                        0
ClaimWihoutIdentifiedThirdParty        0
ClaimAmount                            0
LossHour                              94
PolicyHolderAge                       36
NumberOfBodilyInjuries                 0
FirstPartyLiabil

In [5]:
# Check the unique values for each column
df_insurance_claims.nunique()

ReferenceId                        11530
PolicyholderNumber                  9404
FirstPartyVehicleNumber             9431
ThirdPartyVehicleNumber              350
InsurerNotes                        9040
PolicyholderOccupation                11
LossDate                             804
FirstPolicySubscriptionDate         1389
ClaimCause                            18
ClaimInvolvedCovers                   68
DamageImportance                       2
FirstPartyVehicleType                  4
ConnectionBetweenParties               5
LossPostCode                        3004
PolicyHolderPostCode                3130
PolicyWasSubscribedOnInternet          2
NumberOfPoliciesOfPolicyholder        17
FpVehicleAgeMonths                   388
EasinessToStage                        4
ClaimWihoutIdentifiedThirdParty        2
ClaimAmount                         8248
LossHour                              24
PolicyHolderAge                       73
NumberOfBodilyInjuries                 6
FirstPartyLiabil

#### For the model 1, we want to find out which claims are suspicious if the incident occurs within t days of subscription date. Therefore, we only need columns:
- FirstPolicySubscriptionDate
- LossDate
- Fraud

In [6]:
df_insurance_claims_model_1 = df_insurance_claims[["LossDate", "FirstPolicySubscriptionDate", "Fraud"]]
df_insurance_claims_model_1

LossDate FirstPolicySubscriptionDate  Fraud
0       1/2/19                     6/18/18      0
1       1/2/19                     6/29/17      0
2       1/2/19                      2/5/17      0
3       1/2/19                     1/21/17      0
4       1/2/19                     1/13/18      0
...        ...                         ...    ...
11525  2/17/21                     3/15/19      0
11526   3/7/21                     7/20/17      0
11527  3/15/21                     9/30/20      0
11528   3/6/21                    12/28/18      0
11529  3/13/21                      3/1/17      0

[11530 rows x 3 columns]

#### Based on the "LossDate" and "FirstPolicySubscriptionDate", we calculated the gap days as "Days" between these two columns

In [7]:
df_insurance_claims_model_1["LossDate"] = pd.to_datetime(df_insurance_claims_model_1["LossDate"])
df_insurance_claims_model_1["FirstPolicySubscriptionDate"] = pd.to_datetime(df_insurance_claims_model_1["FirstPolicySubscriptionDate"])

C:\Users\98179\AppData\Local\Temp\ipykernel_25696\1508960330.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_insurance_claims_model_1["LossDate"] = pd.to_datetime(df_insurance_claims_model_1["LossDate"])
C:\Users\98179\AppData\Local\Temp\ipykernel_25696\1508960330.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_insurance_claims_model_1["FirstPolicySubscriptionDate"] = pd.to_datetime(df_insurance_claims_model_1["FirstPolicySubscriptionDate"])


In [8]:
df_insurance_claims_model_1["Days"] = df_insurance_claims_model_1["LossDate"] - df_insurance_claims_model_1["FirstPolicySubscriptionDate"]
df_insurance_claims_model_1["Days"] = pd.to_numeric(df_insurance_claims_model_1["Days"].dt.days, downcast='integer')
df_insurance_claims_model_1 = df_insurance_claims_model_1.astype({"Days": "int64"})

C:\Users\98179\AppData\Local\Temp\ipykernel_25696\3779345626.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_insurance_claims_model_1["Days"] = df_insurance_claims_model_1["LossDate"] - df_insurance_claims_model_1["FirstPolicySubscriptionDate"]
C:\Users\98179\AppData\Local\Temp\ipykernel_25696\3779345626.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_insurance_claims_model_1["Days"] = pd.to_numeric(df_insurance_claims_model_1["Days"].dt.days, downcast='integer')


In [9]:
df_insurance_claims_model_1

LossDate FirstPolicySubscriptionDate  Fraud  Days
0     2019-01-02                  2018-06-18      0   198
1     2019-01-02                  2017-06-29      0   552
2     2019-01-02                  2017-02-05      0   696
3     2019-01-02                  2017-01-21      0   711
4     2019-01-02                  2018-01-13      0   354
...          ...                         ...    ...   ...
11525 2021-02-17                  2019-03-15      0   705
11526 2021-03-07                  2017-07-20      0  1326
11527 2021-03-15                  2020-09-30      0   166
11528 2021-03-06                  2018-12-28      0   799
11529 2021-03-13                  2017-03-01      0  1473

[11530 rows x 4 columns]

In [10]:
df_insurance_claims_model_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11530 entries, 0 to 11529
Data columns (total 4 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   LossDate                     11530 non-null  datetime64[ns]
 1   FirstPolicySubscriptionDate  11530 non-null  datetime64[ns]
 2   Fraud                        11530 non-null  int64         
 3   Days                         11530 non-null  int64         
dtypes: datetime64[ns](2), int64(2)
memory usage: 360.4 KB


#### Determine the range of the t days

In [11]:
df_fraud = df_insurance_claims_model_1.loc[df_insurance_claims_model_1["Fraud"] == 1]

In [12]:
# The maximum value of t
df_fraud["Days"].max()

1201

In [13]:
# The minimum value of t
df_fraud["Days"].min()

0

In [14]:
# Set up the range of t days
t = range(1202)

#### We calculated the detection rate and hit rate based on the each t days
- The calculation formula of detection rate is:
$$
  Detection\ Rate =\frac{The\ Number\ of\ Detection}{Total\ Number\ of\ Observation}\, *100\%
$$
- The calculation formula of hit rate is:
$$
  Hit\ Rate =\frac{The\ Number\ of\ Real\ Fraud}{The\ Number\ of\ Detection}\, *100\%
$$


In [15]:
def detect_fraudulent_claims_model_1(t, df_datasets):
    
    # All_detection rate
    d = []
    # All hit rate
    h = []
    # All predictions 
    Y_all_prediction = []
    
    for i in t:
        # All prediction based on each t
        y_prediction = []
        # The number of detection
        a = 0
        # The number of real fraud
        b = 0
        
        for o in range(len(df_datasets)):
            if df_datasets["Days"][o] <= i:
                a = a + 1
                y = 1
                y_prediction.append(y)
                if df_datasets["Fraud"][o] == 1:
                    b = b + 1
            else:
                y = 0
                y_prediction.append(y)
        
        detection_rate = round((a / len(df_datasets))*100, 2)
        d.append(detection_rate)
        hit_rate = round((b / a)*100, 2)
        h.append(hit_rate)
        
        Y_all_prediction.append(y_prediction)

    
    return d, h, Y_all_prediction

In [ ]:
d, h, Y_all_prediction = detect_fraudulent_claims_model_1(t, df_insurance_claims_model_1)

### The hit rate and detection rate of this model for different values of t are:

#### The hit rate (from t equals 0 to 1201)

In [ ]:
h

#### The detection rate (from t equals 0 to 1201)

In [ ]:
d

### Plot the results of detection rate and hit rate

In [ ]:
fig = plt.figure(figsize = (20, 20))
AX1 = fig.add_subplot(2,2,1)
AX2 = fig.add_subplot(2,2,2)

AX1.plot(t, d, ls = "--", color = "blue")
AX2.plot(t, h, ls = "--", color = "red")

AX1.set_title("The Detection Rate Based on different t Days", fontsize = 15, horizontalalignment="center")
AX2.set_title("The Hit Rate Based on different t Days", fontsize = 15, horizontalalignment="center")

AX1.set_xlabel("Days", fontsize = 10)
AX1.set_ylabel("Detection Rate", fontsize = 10)

AX2.set_xlabel("Days", fontsize = 10)
AX2.set_ylabel("Hit Rate", fontsize = 10)

In order to choose a sensible value of t, we used Receiver Operating Characteristic(ROC) accuracy to evaluate which t is the sensible one. 

A large ROC accuracy indicates that this model with a specific t is able to accurately identify suspicious claims while minimizing the number of false positives. 

- The calculation formula of ROC accuracy is:

$$
  ROC\ Accuracy =\frac{TP\ +\ TN}{TP\ +\ TN\ +\ FP\ +\ FN}
$$


In [ ]:
def get_all_roc_accuracy(y, y_hat):
    all_roc_acc = []
    for i in range(len(t)):
        roc_acc = round(metrics.roc_auc_score(y, y_hat[i]), 2)
        all_roc_acc.append(roc_acc)
    return all_roc_acc

In [ ]:
all_roc_acc = get_all_roc_accuracy(df_insurance_claims_model_1["Fraud"], Y_all_prediction)

#### The ROC accuracy of the model (from t equals 0 to 1201)

In [ ]:
all_roc_acc

### Plot the ROC accuracy based on each t days

In [ ]:
fig = plt.figure(figsize = (15, 5))
AX1 = fig.add_subplot(1,1,1)

AX1.plot(t, all_roc_acc, ls = "-", color = "green")

AX1.set_title("The ROC accuracy Based on different t Days", fontsize = 15, horizontalalignment="center")

AX1.set_xlabel("t days", fontsize = 10)
AX1.set_ylabel("ROC accuracy", fontsize = 10)


### Find out the sensible t of this model

In [ ]:
for i in range(len(all_roc_acc)):
    if all_roc_acc[i] == max(all_roc_acc):
        print("The sensible t is " + str(i))
        print("The detection rate is " + str(d[i]) + "%")
        print("The hit rate is " + str(h[i]) + "%")
        print("The ROC accuracy of the model is " + str(all_roc_acc[i]))

### The logic underlies this model and its bias

The logic underlying this model is that when a customer subscribes to a service, it is more likely that fraudulent claims will be made shortly after the subscription date. Therefore, by setting a threshold for how soon a claim can be made after the subscription date, it is possible to identify suspicious claims. This threshold can then be adjusted to find a balance between the hit rate and detection rate for the model.


However, this model has the potential to discriminate against customers who make claims shortly after subscribing to a service, as these claims will be flagged as suspicious regardless of their validity. That would damage the customers' trust relationship with the company. Additionally, if the threshold is set too low, then the model may also falsely flag legitimate claims as suspicious.